### DISCUSSION TITLE
Can I use `Microsoft.CodeAnalysis.BannedApiAnalyzers` as a library to analyze source code in a folder?

### DISCUSSION THREAD

**Discussion by MartyIX** [dotnet/roslyn#74974] [Q&A] [answered]
I find `BannedApiAnalyzers` ([doc](https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.BannedApiAnalyzers/BannedApiAnalyzers.Help.md)) to be a robust analyzer for checking if source code contain forbidden API or not. I would like to use it _as a library_ in the following way (pseudocode):

```cs
var results = BannedApiAnalyzer.Analyze(files, bannedSymbolsTxt);
```

My question is twofold:

1. Can I do it?
2. Should I actually do it or is it not a supported use case?

Thank you!

**Comment by 333fred** [comment, marked as answer]
You cannot do this, it is not a supported use case. What you _could_ do is use Roslyn itself to create a compilation for those files, add the banned api analyzer to that compilation, and then get all diagnostics.

**Reply by MartyIX** [reply]
I converged to the same approach as here: https://stackoverflow.com/questions/77051745/how-to-use-banned-api-analyzers-in-own-code to find out (as the guy) that `CSharpSymbolIsBannedAnalyzer` is unknown type.

So I guess it's not exported somehow. Or how can one overcome this?

**Reply by MartyIX** [reply]
The type itself is `public`: https://github.com/dotnet/roslyn-analyzers/blob/372f0d4dd91199e415482a1ee302ee27162b885b/src/Microsoft.CodeAnalysis.BannedApiAnalyzers/CSharp/CSharpSymbolIsBannedAnalyzer.cs#L13C25-L13C54 So I guess that it's not exported by its nuget package or something.

**Reply by 333fred** [reply]
You'd have to actually show your project. At a guess, you're depending on the package as an _analyzer_, not as a reference; you'll likely need to do a bit of msbuild magic to get a proper reference.

**Reply by MartyIX** [reply]
I created a minimum repro project here https://github.com/MartyIX/BannedApiRepro. This line https://github.com/MartyIX/BannedApiRepro/blob/ee700c9624154a645af55dcccaf684bac792e7a4/Program.cs#L7 reports a build error because `CSharpSymbolIsBannedAnalyzer`[^1] is not available

[^1]: https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.BannedApiAnalyzers/CSharp/CSharpSymbolIsBannedAnalyzer.cs

**Reply by 333fred** [reply]
Yes, that's what I expected. As I mentioned, you're depending on the analyzer as an analyzer, not a reference. You'll need to do some msbuild logic to change that.

**Reply by MartyIX** [reply]
Thank you. Any idea what to look for specifically? 

I guess I can repost the question to the msbuild repo. They might know the answer.
